Nginx 1.9.11 ngx_udp_connection_t => ngx_resolver_connection_t

charlesportwoodii opened the pull request:
Hi,

In Nginx 1.9.11, the ```ngx_udp_connection_t``` structure has been renamed to ```ngx_resolver_connection_t``` in ```src/core/ngx_resolver.h``` of Nginx core. (https://github.com/nginx/nginx/commit/1945fff7bfee1eef23933faa0ccfbe05445dcfe1).

Consequently, compilation fails in ```lua-nginx-module/src/ngx_http_lua_socket_udp.h``` and ```lua-nginx-module/src/ngx_http_lua_socket_udp.c``` due to references of ```ngx_udp_connection_t```. The compilation failure is as follows:

```
modules/lua-nginx-module/src/ngx_http_lua_util.c
In file included from modules/lua-nginx-module/src/ngx_http_lua_util.c:38:0:
modules/lua-nginx-module/src/ngx_http_lua_socket_udp.h:34:5: error: unknown type name ‘ngx_udp_connection_t’
     ngx_udp_connection_t             udp_connection;
```

I've attached a simple patch that corrects the issue and allows Nginx 1.9.11 to compile.

----
UPDATE: See #668 for others reporting this issue.

nzt4567 commented:
Great work!

agentzh commented on src/ngx_http_lua_socket_udp.c:
```diff
@@ -170,7 +170,7 @@ ngx_http_lua_socket_udp_setpeername(lua_State *L)
     ngx_int_t                    rc;
     ngx_http_lua_loc_conf_t     *llcf;
-    ngx_udp_connection_t        *uc;
+    ngx_resolver_connection_t    *uc;
```
Style: the identifier names should be aligned up vertically. I think the `ngx-releng` tool in the nginx-devel-utils repos automatically catches coding style issues like this:

https://github.com/openresty/nginx-devel-utils

Will you please run this tool and fix any coding style issues? Thanks!

agentzh commented on src/ngx_http_lua_socket_udp.h:
```diff
@@ -31,7 +31,7 @@ struct ngx_http_lua_socket_udp_upstream_s {
     ngx_http_cleanup_pt             *cleanup;
     ngx_http_request_t              *request;
-    ngx_udp_connection_t             udp_connection;
+    ngx_resolver_connection_t       connection;
```
I think we should still keep the original `udp_connection` field name here. Also the identifier is not aligned up properly.

agentzh commented on src/ngx_http_lua_socket_udp.c:
```diff
@@ -54,7 +54,7 @@ static void ngx_http_lua_socket_udp_read_handler(ngx_http_request_t *r,
 static void ngx_http_lua_socket_udp_handle_success(ngx_http_request_t *r,
     ngx_http_lua_socket_udp_upstream_t *u);
-static ngx_int_t ngx_http_lua_udp_connect(ngx_udp_connection_t *uc);
+static ngx_int_t ngx_http_lua_udp_connect(ngx_resolver_connection_t *uc);
```
This change will break builds with nginx cores older than 1.9.11, I'm afraid. We need macros to support both.

agentzh replied:
A closer look at the definition of `ngx_resolver_connection_t` reveals that this type does not serve our purpose well since it contains unused fields like `read_buf` and `write_buf`. Maybe it's time to roll out our own `ngx_http_lua_udp_connection_t` type instead?

charlesportwoodii commented:
@agentzh,

Sorry about the style. I've rebased an update that has the alignment issues correct.

Changes in this rebase:
- Changed ```connection``` back to ```udp_connection```
- Styles

I'm not exactly sure what should be done regarding ```ngx_resolver_connection_t```. In Nginx core, ```ngx_udp_connection_t``` was simply renamed to ```ngx_resolver_connection_t``` with some slight variation to the struct (```connection``` was renamed to ```udp```). 

It may be beneficial to to create something like a ```ngx_http_lua_udp_connection_t``` struct, as that would be an easier place to put the macros in. My C is a little bit rusty though, and I just whipped to get Nginx to compile again = ).

agentzh commented:
@charlesportwoodii Thanks for your work on this. I'll continue thinking about the details here.

charlesportwoodii commented:
@agentzh,

Glad to help out what little I can. Sorry for not being much more useful than this though. Preprocessor macros will be a pretty simple solution for the backwards compatibility aspect, but that would make a mess of ```src/ngx_http_lua_socket_udp.c``` and I wanted to avoid that as much as possible.

Let me know if I can help out in any other way.

agentzh commented:
@charlesportwoodii Thanks for your help thus far :) I'll look into this in the next couple of weeks. The next OpenResty release will be based on NGINX 1.9.11 anyway.

agentzh commented:
The compilation error should be gone now. Thanks for the report!
